Flaky test: TestRootRotationReconciliationNoChanges

I saw this failure in CI (https://circleci.com/gh/docker/swarmkit/7002). It does not seem to be related to the changes in that PR.

```
--- FAIL: TestRootRotationReconciliationNoChanges (3.05s)
	Error Trace:	server_test.go:1108
	Error:		Not equal: []byte{0xa, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0x42, 0x45, 0x47, 0x49, 0x4e, 0x20, 0x45, 0x43, 0x20, 0x50, 0x52, 0x49, 0x56, 0x41, 0x54, 0x45, 0x20, 0x4b, 0x45, 0x59, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0xa, 0x4d, 0x48, 0x63, 0x43, 0x41, 0x51, 0x45, 0x45, 0x49, 0x4b, 0x58, 0x6b, 0x76, 0x46, 0x66, 0x55, 0x63, 0x56, 0x62, 0x48, 0x39, 0x55, 0x71, 0x78, 0x6b, 0x64, 0x6f, 0x34, 0x4f, 0x62, 0x77, 0x63, 0x33, 0x52, 0x53, 0x4a, 0x66, 0x45, 0x48, 0x32, 0x32, 0x35, 0x34, 0x73, 0x66, 0x71, 0x6b, 0x78, 0x35, 0x30, 0x78, 0x42, 0x6f, 0x41, 0x6f, 0x47, 0x43, 0x43, 0x71, 0x47, 0x53, 0x4d, 0x34, 0x39, 0xa, 0x41, 0x77, 0x45, 0x48, 0x6f, 0x55, 0x51, 0x44, 0x51, 0x67, 0x41, 0x45, 0x65, 0x62, 0x4a, 0x2b, 0x41, 0x55, 0x6b, 0x75, 0x37, 0x33, 0x67, 0x6a, 0x49, 0x39, 0x68, 0x35, 0x69, 0x2f, 0x2b, 0x56, 0x6f, 0x4e, 0x52, 0x37, 0x70, 0x78, 0x64, 0x78, 0x6c, 0x5a, 0x6b, 0x76, 0x72, 0x5a, 0x31, 0x62, 0x65, 0x32, 0x62, 0x72, 0x56, 0x68, 0x51, 0x4e, 0x46, 0x42, 0x76, 0x6e, 0x76, 0x4d, 0x6e, 0x47, 0xa, 0x39, 0x6f, 0x68, 0x34, 0x65, 0x78, 0x33, 0x73, 0x70, 0x62, 0x52, 0x6a, 0x42, 0x41, 0x52, 0x73, 0x35, 0x36, 0x48, 0x5a, 0x43, 0x6a, 0x52, 0x52, 0x70, 0x76, 0x6a, 0x36, 0x6d, 0x42, 0x4f, 0x79, 0x4a, 0x51, 0x3d, 0x3d, 0xa, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0x45, 0x4e, 0x44, 0x20, 0x45, 0x43, 0x20, 0x50, 0x52, 0x49, 0x56, 0x41, 0x54, 0x45, 0x20, 0x4b, 0x45, 0x59, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0xa} (expected)
			        != []byte{0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0x42, 0x45, 0x47, 0x49, 0x4e, 0x20, 0x45, 0x43, 0x20, 0x50, 0x52, 0x49, 0x56, 0x41, 0x54, 0x45, 0x20, 0x4b, 0x45, 0x59, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0xa, 0x4d, 0x48, 0x63, 0x43, 0x41, 0x51, 0x45, 0x45, 0x49, 0x50, 0x76, 0x32, 0x71, 0x63, 0x46, 0x6f, 0x4f, 0x79, 0x34, 0x65, 0x6e, 0x4a, 0x59, 0x2b, 0x45, 0x39, 0x7a, 0x62, 0x62, 0x7a, 0x63, 0x4b, 0x72, 0x5a, 0x4d, 0x6c, 0x73, 0x34, 0x43, 0x44, 0x48, 0x77, 0x62, 0x6e, 0x2f, 0x38, 0x49, 0x74, 0x61, 0x63, 0x79, 0x73, 0x6f, 0x41, 0x6f, 0x47, 0x43, 0x43, 0x71, 0x47, 0x53, 0x4d, 0x34, 0x39, 0xa, 0x41, 0x77, 0x45, 0x48, 0x6f, 0x55, 0x51, 0x44, 0x51, 0x67, 0x41, 0x45, 0x48, 0x54, 0x34, 0x63, 0x6e, 0x55, 0x4d, 0x44, 0x58, 0x7a, 0x33, 0x58, 0x6b, 0x68, 0x37, 0x6c, 0x57, 0x68, 0x49, 0x2b, 0x53, 0x47, 0x56, 0x57, 0x36, 0x4b, 0x41, 0x74, 0x6d, 0x7a, 0x64, 0x76, 0x32, 0x6a, 0x42, 0x6d, 0x42, 0x57, 0x6d, 0x5a, 0x52, 0x32, 0x34, 0x4f, 0x44, 0x77, 0x38, 0x4c, 0x6e, 0x35, 0x44, 0x4b, 0xa, 0x48, 0x56, 0x43, 0x5a, 0x32, 0x67, 0x6d, 0x4c, 0x73, 0x36, 0x67, 0x69, 0x4b, 0x48, 0x73, 0x31, 0x69, 0x31, 0x6a, 0x36, 0x56, 0x49, 0x69, 0x55, 0x62, 0x4b, 0x30, 0x49, 0x31, 0x79, 0x55, 0x45, 0x47, 0x77, 0x3d, 0x3d, 0xa, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0x45, 0x4e, 0x44, 0x20, 0x45, 0x43, 0x20, 0x50, 0x52, 0x49, 0x56, 0x41, 0x54, 0x45, 0x20, 0x4b, 0x45, 0x59, 0x2d, 0x2d, 0x2d, 0x2d, 0x2d, 0xa} (actual)
			
			Diff:
			--- Expected
			+++ Actual
			@@ -1,17 +1,17 @@
			-([]uint8) (len=228 cap=228) {
			- 00000000  0a 2d 2d 2d 2d 2d 42 45  47 49 4e 20 45 43 20 50  |.-----BEGIN EC P|
			- 00000010  52 49 56 41 54 45 20 4b  45 59 2d 2d 2d 2d 2d 0a  |RIVATE KEY-----.|
			- 00000020  4d 48 63 43 41 51 45 45  49 4b 58 6b 76 46 66 55  |MHcCAQEEIKXkvFfU|
			- 00000030  63 56 62 48 39 55 71 78  6b 64 6f 34 4f 62 77 63  |cVbH9Uqxkdo4Obwc|
			- 00000040  33 52 53 4a 66 45 48 32  32 35 34 73 66 71 6b 78  |3RSJfEH2254sfqkx|
			- 00000050  35 30 78 42 6f 41 6f 47  43 43 71 47 53 4d 34 39  |50xBoAoGCCqGSM49|
			- 00000060  0a 41 77 45 48 6f 55 51  44 51 67 41 45 65 62 4a  |.AwEHoUQDQgAEebJ|
			- 00000070  2b 41 55 6b 75 37 33 67  6a 49 39 68 35 69 2f 2b  |+AUku73gjI9h5i/+|
			- 00000080  56 6f 4e 52 37 70 78 64  78 6c 5a 6b 76 72 5a 31  |VoNR7pxdxlZkvrZ1|
			- 00000090  62 65 32 62 72 56 68 51  4e 46 42 76 6e 76 4d 6e  |be2brVhQNFBvnvMn|
			- 000000a0  47 0a 39 6f 68 34 65 78  33 73 70 62 52 6a 42 41  |G.9oh4ex3spbRjBA|
			- 000000b0  52 73 35 36 48 5a 43 6a  52 52 70 76 6a 36 6d 42  |Rs56HZCjRRpvj6mB|
			- 000000c0  4f 79 4a 51 3d 3d 0a 2d  2d 2d 2d 2d 45 4e 44 20  |OyJQ==.-----END |
			- 000000d0  45 43 20 50 52 49 56 41  54 45 20 4b 45 59 2d 2d  |EC PRIVATE KEY--|
			- 000000e0  2d 2d 2d 0a                                       |---.|
			+([]uint8) (len=227 cap=227) {
			+ 00000000  2d 2d 2d 2d 2d 42 45 47  49 4e 20 45 43 20 50 52  |-----BEGIN EC PR|
			+ 00000010  49 56 41 54 45 20 4b 45  59 2d 2d 2d 2d 2d 0a 4d  |IVATE KEY-----.M|
			+ 00000020  48 63 43 41 51 45 45 49  50 76 32 71 63 46 6f 4f  |HcCAQEEIPv2qcFoO|
			+ 00000030  79 34 65 6e 4a 59 2b 45  39 7a 62 62 7a 63 4b 72  |y4enJY+E9zbbzcKr|
			+ 00000040  5a 4d 6c 73 34 43 44 48  77 62 6e 2f 38 49 74 61  |ZMls4CDHwbn/8Ita|
			+ 00000050  63 79 73 6f 41 6f 47 43  43 71 47 53 4d 34 39 0a  |cysoAoGCCqGSM49.|
			+ 00000060  41 77 45 48 6f 55 51 44  51 67 41 45 48 54 34 63  |AwEHoUQDQgAEHT4c|
			+ 00000070  6e 55 4d 44 58 7a 33 58  6b 68 37 6c 57 68 49 2b  |nUMDXz3Xkh7lWhI+|
			+ 00000080  53 47 56 57 36 4b 41 74  6d 7a 64 76 32 6a 42 6d  |SGVW6KAtmzdv2jBm|
			+ 00000090  42 57 6d 5a 52 32 34 4f  44 77 38 4c 6e 35 44 4b  |BWmZR24ODw8Ln5DK|
			+ 000000a0  0a 48 56 43 5a 32 67 6d  4c 73 36 67 69 4b 48 73  |.HVCZ2gmLs6giKHs|
			+ 000000b0  31 69 31 6a 36 56 49 69  55 62 4b 30 49 31 79 55  |1i1j6VIiUbK0I1yU|
			+ 000000c0  45 47 77 3d 3d 0a 2d 2d  2d 2d 2d 45 4e 44 20 45  |EGw==.-----END E|
			+ 000000d0  43 20 50 52 49 56 41 54  45 20 4b 45 59 2d 2d 2d  |C PRIVATE KEY---|
			+ 000000e0  2d 2d 0a                                          |--.|
			 }
	Messages:	Nodes already in rotate state, even if they currently have the correct TLS issuer, will be left in the rotate state even if root rotation is aborted because we don't know if they're already in the process of getting a new cert.  Even if they're issued by a different issuer, they will be left alone because they'll have an interemdiate that chains up to the old issuer.
```

--- a/ca/reconciler.go
+++ b/ca/reconciler.go
@@ -99,7 +99,7 @@ func (r *rootRotationReconciler) UpdateRootCA(newRootCA *api.RootCA) {
 	if newRootCA.RootRotation != nil {
 		var nodes []*api.Node
 		r.store.View(func(tx store.ReadTx) {
-			nodes, err = store.FindNodes(tx, store.ByMembership(api.NodeMembershipAccepted))
+			nodes, err = store.FindNodes(tx, store.All)
 		})
 		if err != nil {
 			log.G(r.ctx).WithError(err).Error("unable to list nodes, so unable to process the current root CA")
@@ -132,8 +132,8 @@ func (r *rootRotationReconciler) UpdateRootCA(newRootCA *api.RootCA) {
 func (r *rootRotationReconciler) UpdateNode(node *api.Node) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	// if we're not in the middle of a root rotation, or if this node does not have membership, ignore it
-	if r.currentRootCA == nil || r.currentRootCA.RootRotation == nil || node.Spec.Membership != api.NodeMembershipAccepted {
+	// if we're not in the middle of a root rotation ignore the update
+	if r.currentRootCA == nil || r.currentRootCA.RootRotation == nil {
 		return
 	}
 	if hasIssuer(node, &r.currentIssuer) {

--- a/ca/server_test.go
+++ b/ca/server_test.go
@@ -725,6 +725,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, nil, true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStatePending, nil, true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateFailed, nil, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, nil, false),
 			},
 			rootCA: &api.RootCA{
 				CACert:     startCluster.RootCA.CACert,
@@ -743,6 +744,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, nil, true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateRotate, nil, true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateRotate, nil, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateRotate, nil, false),
 			},
 		},
 		{
@@ -756,6 +758,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateIssued, oldNodeTLSInfo, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, oldNodeTLSInfo, false),
 			},
 			rootCA: &api.RootCA{ // no change in root CA from previous
 				CACert:     startCluster.RootCA.CACert,
@@ -774,6 +777,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateRotate, oldNodeTLSInfo, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateRotate, oldNodeTLSInfo, false),
 			},
 		},
 		{
@@ -784,7 +788,8 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateIssued, rotationTLSInfo[0], true),
-				"6": getFakeAPINode(t, "6", api.IssuanceStateRenew, nil, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, rotationTLSInfo[0], false),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateRenew, nil, true),
 			},
 			rootCA: &api.RootCA{ // no change in root CA from previous
 				CACert:     startCluster.RootCA.CACert,
@@ -802,7 +807,8 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateIssued, rotationTLSInfo[0], true),
-				"6": getFakeAPINode(t, "6", api.IssuanceStateRotate, nil, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, rotationTLSInfo[0], false),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateRotate, nil, true),
 			},
 		},
 		{
@@ -816,6 +822,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateIssued, oldNodeTLSInfo, true),
 				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, rotationTLSInfo[0], true),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateIssued, rotationTLSInfo[0], false),
 			},
 			rootCA: &api.RootCA{ // new root rotation
 				CACert:     startCluster.RootCA.CACert,
@@ -834,13 +841,14 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"4": getFakeAPINode(t, "4", api.IssuanceStateRotate, rotationTLSInfo[0], true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateRotate, oldNodeTLSInfo, true),
 				"6": getFakeAPINode(t, "6", api.IssuanceStateRotate, rotationTLSInfo[0], true),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateRotate, rotationTLSInfo[0], false),
 			},
 		},
 		{
 			descr: ("Once all nodes have rotated to their desired TLS info (even if it's because " +
 				"a node with the wrong TLS info has been removed, the root rotation is completed."),
 			nodes: map[string]*api.Node{
-				"0": getFakeAPINode(t, "0", api.IssuanceStatePending, nil, false),
+				"0": getFakeAPINode(t, "0", api.IssuanceStateIssued, rotationTLSInfo[1], false),
 				"1": getFakeAPINode(t, "1", api.IssuanceStateIssued, rotationTLSInfo[1], true),
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[1], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[1], true),
@@ -877,6 +885,7 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateIssued, rotationTLSInfo[1], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateIssued, rotationTLSInfo[1], true),
 				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, rotationTLSInfo[1], true),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateIssued, rotationTLSInfo[1], false),
 			},
 			rootCA: &api.RootCA{
 				CACert:     startCluster.RootCA.CACert,
@@ -894,27 +903,33 @@ func TestRootRotationReconciliationWithChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, rotationTLSInfo[1], true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStateRotate, rotationTLSInfo[1], true),
 				"6": getFakeAPINode(t, "6", api.IssuanceStateRotate, rotationTLSInfo[1], true),
+				"7": getFakeAPINode(t, "7", api.IssuanceStateRotate, rotationTLSInfo[1], false),
 			},
 		},
 	}
 
 	for _, testcase := range testcases {
-		if testcase.caServerRestart {
-			rt.tc.CAServer.Stop()
-		}
-
-		rt.convergeRootCA(testcase.rootCA, testcase.descr)
+		// stop the CA server, get the cluster to the state we want (correct root CA, correct nodes, etc.)
+		rt.tc.CAServer.Stop()
 		rt.convergeWantedNodes(testcase.nodes, testcase.descr)
+
+		if testcase.caServerRestart {
+			// if we want to simulate restarting the CA server with a root rotation already done, set the rootCA to
+			// have a root rotation, then start the CA
+			rt.convergeRootCA(testcase.rootCA, testcase.descr)
+			startCAServer(rt.tc.CAServer)
+		} else {
+			// otherwise, start the CA in the state where there is no root rotation, and start a root rotation
+			rt.convergeRootCA(&startCluster.RootCA, testcase.descr) // no root rotation
+			startCAServer(rt.tc.CAServer)
+			rt.convergeRootCA(testcase.rootCA, testcase.descr)
+		}
 
 		if testcase.expectedNodes == nil {
 			testcase.expectedNodes = testcase.nodes
 		}
 		if testcase.expectedRootCA == nil {
 			testcase.expectedRootCA = testcase.rootCA
-		}
-
-		if testcase.caServerRestart {
-			startCAServer(rt.tc.CAServer)
 		}
 
 		require.NoError(t, testutils.PollFuncWithTimeout(nil, func() error {
@@ -1022,6 +1037,7 @@ func TestRootRotationReconciliationNoChanges(t *testing.T) {
 				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, nil, true),
 				"4": getFakeAPINode(t, "4", api.IssuanceStatePending, nil, true),
 				"5": getFakeAPINode(t, "5", api.IssuanceStateFailed, nil, true),
+				"6": getFakeAPINode(t, "6", api.IssuanceStateIssued, oldNodeTLSInfo, false),
 			},
 			rootCA: &api.RootCA{
 				CACert:     startCluster.RootCA.CACert,
@@ -1035,13 +1051,13 @@ func TestRootRotationReconciliationNoChanges(t *testing.T) {
 			},
 		},
 		{
-			descr: ("If all nodes have the right TLS info or are already rotated (or are not members), " +
+			descr: ("If all nodes have the right TLS info or are already rotated, rotating, or pending, " +
 				"there will be no changes needed"),
 			nodes: map[string]*api.Node{
 				"0": getFakeAPINode(t, "0", api.IssuanceStatePending, nil, false),
 				"1": getFakeAPINode(t, "1", api.IssuanceStateIssued, rotationTLSInfo, true),
 				"2": getFakeAPINode(t, "2", api.IssuanceStateRotate, oldNodeTLSInfo, true),
-				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, rotationTLSInfo, true),
+				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, rotationTLSInfo, false),
 			},
 			rootCA: &api.RootCA{ // no change in root CA from previous
 				CACert:     startCluster.RootCA.CACert,
@@ -1063,6 +1079,7 @@ func TestRootRotationReconciliationNoChanges(t *testing.T) {
 				"0": getFakeAPINode(t, "0", api.IssuanceStatePending, nil, false),
 				"1": getFakeAPINode(t, "1", api.IssuanceStateIssued, rotationTLSInfo, true),
 				"2": getFakeAPINode(t, "2", api.IssuanceStateRotate, oldNodeTLSInfo, true),
+				"3": getFakeAPINode(t, "3", api.IssuanceStateRotate, oldNodeTLSInfo, false),
 			},
 			rootCA: &api.RootCA{ // no change in root CA from previous
 				CACert:     startCluster.RootCA.CACert,
@@ -1073,14 +1090,15 @@ func TestRootRotationReconciliationNoChanges(t *testing.T) {
 	}
 
 	for _, testcase := range testcases {
-		if testcase.caServerStopped {
-			rt.tc.CAServer.Stop()
-		} else {
+		// stop the CA server, get the cluster to the state we want (correct root CA, correct nodes, etc.)
+		rt.tc.CAServer.Stop()
+		rt.convergeWantedNodes(testcase.nodes, testcase.descr)
+		rt.convergeRootCA(&startCluster.RootCA, testcase.descr) // no root rotation
+
+		if !testcase.caServerStopped {
 			startCAServer(rt.tc.CAServer)
 		}
-
 		rt.convergeRootCA(testcase.rootCA, testcase.descr)
-		rt.convergeWantedNodes(testcase.nodes, testcase.descr)
 
 		time.Sleep(500 * time.Millisecond)
 
@@ -1286,7 +1304,7 @@ func TestRootRotationReconciliationThrottled(t *testing.T) {
 		tc.MemoryStore, func(tx store.ReadTx) error {
 			// don't bother getting the cluster - the CA server has already done that when first running
 			var err error
-			nodes, err = store.FindNodes(tx, store.ByMembership(api.NodeMembershipAccepted))
+			nodes, err = store.FindNodes(tx, store.All)
 			return err
 		},
 		api.EventUpdateCluster{

